### PR TITLE
Implement QUIC address validation

### PR DESCRIPTION
**Pull request by nhorman**
QUIC server address verification is a requirement for QUIC server MVP.  This pull request, based initially on @Sashan work implements that feature and fixes the various issues that were left to solve on it, as well as the various tests that were impacted by its introduction

##### Checklist
- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by nhorman** [comment]
Note: Waiving style due to quic_connection.h header having consistent out-of-line style

**Comment by nhorman** [comment]
CI failures seem not relevant, as they are failing tests for cloudflare validating the kyber algorithm

File: ssl/quic/quic_channel.c [hunk 1]
```diff
@@ -1238,36 +1237,6 @@ static int ch_on_transport_params(const unsigned char *params,
...
-
-            /* Must match Retry packet SCID. */
-            if (!ossl_quic_conn_id_eq(&ch->retry_scid, &cid)) {
-                reason = TP_REASON_EXPECTED_VALUE("RETRY_SCID");
-                goto malformed;
-            }
-
-            got_retry_scid = 1;
-            break;
-
```
Comment (t8m) [inline ssl/quic/quic_channel.c:1270]: Why did you remove this? This does not look right. Please see https://www.ietf.org/rfc/rfc9000.html#cid-auth

**Review by t8m — CHANGES_REQUESTED** [review]

File: ssl/quic/quic_channel.c (hunk 1 above)
Comment (Sashan) [reply ssl/quic/quic_channel.c:1270]: I agree, this removal makes PR too invasive. I prefer to keep it.

File: include/internal/quic_channel.h [hunk 2]
```diff
@@ -448,6 +448,10 @@ uint64_t ossl_quic_channel_get_max_idle_timeout_peer_request(const QUIC_CHANNEL
 /* Get the idle timeout actually negotiated. */
 uint64_t ossl_quic_channel_get_max_idle_timeout_actual(const QUIC_CHANNEL *ch);
 
+int ossl_quic_bind_channel(QUIC_CHANNEL *ch, const BIO_ADDR *peer,
```
Comment (Sashan) [inline include/internal/quic_channel.h:451]: I think function deserves a comment. This is my suggestion:
```
+/*
+ * Bind channel instance connection ids, used when server validates client
+ * with retry packet. In this case we know connection ids (CIDs) before
+ * channel is created.
+ */
```

File: include/internal/quic_wire.h [hunk 3]
```diff
@@ -121,7 +121,6 @@ ossl_quic_frame_type_is_ack_eliciting(uint64_t frame_type)
 #  define QUIC_TPARAM_PREFERRED_ADDR                      0x0D
 #  define QUIC_TPARAM_ACTIVE_CONN_ID_LIMIT                0x0E
 #  define QUIC_TPARAM_INITIAL_SCID                        0x0F
-#  define QUIC_TPARAM_RETRY_SCID                          0x10
```
Comment (Sashan) [inline include/internal/quic_wire.h:124]: I think this should stay.

File: ssl/quic/quic_channel.c [hunk 4]
```diff
@@ -1613,11 +1582,6 @@ static int ch_on_transport_params(const unsigned char *params,
             reason = TP_REASON_REQUIRED("ORIG_DCID");
             goto malformed;
         }
```
Comment (Sashan) [inline ssl/quic/quic_channel.c:1584]: Thinking more about it. It looks like our quic server does not include that `retry_source_connection_id` parameter after it sends its initial packet after validating client successfully. it feels like yet another piece which is missing in mosaic.

So I think this piece should also stay.

It looks like that once server creates channel after validating client, the server must remember to include this parameter to pass retry scid to client.

File: ssl/quic/quic_port.c [hunk 5]
```diff
@@ -686,18 +862,29 @@ static void port_default_packet_handler(QUIC_URXE *e, void *arg,
...
+    } else {
+        port_bind_channel(port, &e->peer, &hdr.src_conn_id, &hdr.dst_conn_id,
+                          &odcid, &new_ch);
+    }
+
+    /*
      * The channel will do all the LCID registration needed, but as an
      * optimization inject this packet directly into the channel's QRX for
      * processing without going through the DEMUX again.
      */
```
Comment (Sashan) [inline ssl/quic/quic_port.c:883]: this is captain obvious speaking: the `port_on_new_conn()` becomes currently unused function. but it will come back once we will have answer for TODO note few lines above.

**Review by Sashan — CHANGES_REQUESTED** [review]
nice work Niel. things are heading in right direction. However I think there must be better plan for  `QUIC_TPARAM_RETRY_SCID`

thanks

File: ssl/quic/quic_port.c (hunk 5 above)
Comment (t8m) [reply ssl/quic/quic_port.c:883]: You can just use `#if 0`

File: include/internal/quic_channel.h (hunk 2 above)
Comment (nhorman) [reply include/internal/quic_channel.h:451]: added a comment to the funciton definition, which is where I think all internal functions need to be documented per coding style guide

**Comment by nhorman** [comment]
ok, I've restored all the RETRY_SCID code, and instead made the server include the RETRY_SCID transport param when we establish a connection via retry, which seems to be in keeping with the RFC.  That should address all the comments you both made.

File: test/recipes/75-test_quicapi_data/ssltraceref.txt [hunk 6]
```diff
@@ -79,10 +79,26 @@ Sent Packet
   Packet Number: 0x00000000
 Sent Datagram
   Length: 1200
```
Comment (Sashan) [inline test/recipes/75-test_quicapi_data/ssltraceref.txt:81]: just a question: where do those changes coming from?

**Review by Sashan — CHANGES_REQUESTED** [review]
There is just one nit in comment. otherwise looks good. I think those changes should go in.

thanks.

File: test/recipes/75-test_quicapi_data/ssltraceref.txt (hunk 6 above)
Comment (nhorman) [reply test/recipes/75-test_quicapi_data/ssltraceref.txt:81]: The quicapi test has a subtest for confirming that the SSL_trace api works properly for them.  It does this by hook SSL_trace to trace a QUIC connection, which it then compares against a known trace (which I'm guessing was just recorded long ago).  With the addition of the retry request, the SSL_trace command records new SSL records (which include both the retry frame being sent, as well as additional transport parameters).  This alters the trace records, causing failures in the test.  These fixups bring the expected SSL trace records back into line with what we actually record.

**Review by Sashan — APPROVED** [review]
thank you. looks good to me. It should go in.
nice work.

**Comment by Sashan** [comment]
One more thing: please make sure commit message start with 'Implement QUIC address validation' (or something similar) the key is to keep phrase _address validation_ in so we are consistent with RFC (and rest of the QUIC world). thanks.

otherwise my approval still holds.

**Review by andrewkdinh — CHANGES_REQUESTED** [review]
Some minor fixes

File: ssl/quic/quic_lcidm.c [hunk 7]
```diff
@@ -554,3 +584,16 @@ int ossl_quic_lcidm_debug_add(QUIC_LCIDM *lcidm, void *opaque,
...
     return 1;
 }
+
+int ossl_quic_lcidm_get_unused_cid(QUIC_LCIDM *lcidm, QUIC_CONN_ID *cid)
+{
+    int i;
+
+    for (i = 0; i < 10; i++) {
+        if (lcidm_generate_cid(lcidm, cid) &&
+            (lcidm_get0_lcid(lcidm, cid) == NULL))
```
Comment (FdaSilvaYY) [inline ssl/quic/quic_lcidm.c:594]: useless parenthesis pair here

**Review by FdaSilvaYY — COMMENTED** [review]
very quick fly-over

File: ssl/quic/quic_lcidm.c (hunk 7 above)
Comment (nhorman) [reply ssl/quic/quic_lcidm.c:594]: I think the added parenthesis makes this more understandable

File: ssl/quic/quic_channel.c [hunk 8]
```diff
@@ -1734,12 +1749,17 @@ static int ch_generate_transport_params(QUIC_CHANNEL *ch)
...
+                                                       id_to_use))
             goto err;
 
         if (!ossl_quic_wire_encode_transport_param_cid(&wpkt, QUIC_TPARAM_INITIAL_SCID,
                                                        &ch->cur_local_cid))
             goto err;
+        if (ch->odcid.id_len != 0)
+            if (!ossl_quic_wire_encode_transport_param_cid(&wpkt,
+                                                           QUIC_TPARAM_RETRY_SCID,
+                                                           &ch->init_dcid))
```
Comment (mattcaswell) [inline ssl/quic/quic_channel.c:1761]: We seem to be putting a dcid into a scid transport parameter - which looks suspicious. Is this correct?

File: ssl/quic/quic_channel_local.h [hunk 9]
```diff
@@ -107,6 +107,13 @@ struct quic_channel_st {
      */
```
Comment (mattcaswell) [inline ssl/quic/quic_channel_local.h:107]: Isn't the comment here out of date? It talks about the first Initial packet the client transmitted - but isn't this updated after a retry?

File: ssl/quic/quic_port.c [hunk 10]
```diff
@@ -620,6 +647,154 @@ static int port_try_handle_stateless_reset(QUIC_PORT *port, const QUIC_URXE *e)
...
+    unsigned char buffer[512];
+    WPACKET wpkt;
+    size_t written;
+    QUIC_PKT_HDR hdr;
+    struct validation_token token;
+    size_t token_len = TOKEN_LEN;
+    unsigned char *integrity_tag;
+    int ok;
+
+    /* TODO: generate proper validation token */
```
Comment (mattcaswell) [inline ssl/quic/quic_port.c:690]: TODO(QUIC SERVER)

File: ssl/quic/quic_port.c [hunk 11]
```diff
@@ -620,6 +647,154 @@ static int port_try_handle_stateless_reset(QUIC_PORT *port, const QUIC_URXE *e)
...
+    memcpy(token.token_buf, "openssltoken", sizeof("openssltoken") - 1);
+
+    token.token_odcid = client_hdr->dst_conn_id;
+    token_len += token.token_odcid.id_len;
+    integrity_tag = (unsigned char *)&token.token_odcid +
+        token.token_odcid.id_len + sizeof(token.token_odcid.id_len);
```
Comment (mattcaswell) [inline ssl/quic/quic_port.c:696]: We seem to be treating "token" not as the struct it is defined as but as a byte buffer. This is truly horrible hack. There must be a better way. Better would be to have an actual byte buffer and copy the data into it.

**Review by mattcaswell — CHANGES_REQUESTED** [review]
I started to look at this - but I'm now out of time. Comments so far.

File: ssl/quic/quic_channel_local.h (hunk 9 above)
Comment (nhorman) [reply ssl/quic/quic_channel_local.h:107]: I suppose thats debatable.  When a retry packet is received, the connection sends another Initial packet, which is really the first 'real' Initial packet of the connection, as the previous one is required to be discarded by the server, so you could argue that the init_dcid is still the init_dcid for the connection, as the prior dcid was forgotten

File: ssl/quic/quic_port.c (hunk 11 above)
Comment (nhorman) [reply ssl/quic/quic_port.c:696]: We need to refine how the token construction takes place anyway, in support of its use both in a retry packet and in a NEW_TOKEN frame.  This is being handled in https://github.com/openssl/project/issues/912

File: ssl/quic/quic_lcidm.c (hunk 7 above)
Comment (t8m) [reply ssl/quic/quic_lcidm.c:594]: This is against our coding style and IMO putting `&&` on the beginning of the second line (which needs to be done according to our coding style) makes this understandable very well without them.

File: ssl/quic/quic_port.c [hunk 12]
```diff
@@ -691,18 +867,29 @@ static void port_default_packet_handler(QUIC_URXE *e, void *arg,
         goto undesirable;
 
     /*
-     * Try to process this as a valid attempt to initiate a connection.
-     *
+     * TODO: there should be some logic similar to accounting half-open
```
Comment (t8m) [inline ssl/quic/quic_port.c:870]: Please use `TODO(QUIC SERVER)` instead.

File: ssl/quic/quic_port.c [hunk 13]
```diff
@@ -691,18 +867,29 @@ static void port_default_packet_handler(QUIC_URXE *e, void *arg,
...
+     * states in TCP. If we reach certain threshold, then we want to
+     * validate clients.
+     */
+    if (hdr.token == NULL) {
+        port_send_retry(port, &e->peer, &hdr);
+        goto undesirable;
+    } else if (port_validate_token(&hdr, &odcid) == 0) {
+        goto undesirable;
+    } else {
```
Comment (t8m) [inline ssl/quic/quic_port.c:879]: Nit: this else is unneeded as both branches before do `goto`.

**Review by t8m — CHANGES_REQUESTED** [review]

File: ssl/quic/quic_port.c [hunk 14]
```diff
@@ -620,6 +647,154 @@ static int port_try_handle_stateless_reset(QUIC_PORT *port, const QUIC_URXE *e)
...
+    ok = WPACKET_get_total_written(&wpkt, &msg[0].data_len);
+    if (ok == 0)
+        return;
+
+    ok = WPACKET_finish(&wpkt);
+    if (ok == 0)
+        return;
+
+    if (!BIO_sendmmsg(port->net_wbio, msg, sizeof(BIO_MSG), 1, 0, &written))
+        ERR_raise_data(ERR_LIB_SSL, SSL_R_QUIC_NETWORK_ERROR,
```
Comment (Sashan) [inline ssl/quic/quic_port.c:752]: I've just noticed the same thing at PR which introduces version negotiation. I have a same question here: is it worth to raise error here? If needs to be handled then I would prefer to go with trace/log event.

**Review by Sashan — CHANGES_REQUESTED** [review]
I think we need to agree on how error from BIO_sendmmsg() should be handled. I prefer to just log it instead of doing `ERR_raise_data()` in void function.

File: ssl/quic/quic_port.c (hunk 14 above)
Comment (nhorman) [reply ssl/quic/quic_port.c:752]: A trace event might be better, as the ERR api saves information to the stack of the calling thread, and thats non-deterministic here

Comment (nhorman) [reply ssl/quic/quic_port.c:752]: followup question, should we have a QUIC trace category here?  The closest category we have now is TLS, and that doesn't seem to fit the bill here

Comment (t8m) [reply ssl/quic/quic_port.c:752]: Hmm... this is even more problematic. The bio is non-blocking and it might fail with EAGAIN expecting that you call the sendmmsg() again when it becomes writable. So there is definitely at least TODO(QUIC_SERVER) to resolve here.

Comment (Sashan) [reply ssl/quic/quic_port.c:752]: there is also a qlog subsystem for QUIC. however not sure if there is any mean to signal/log failure in sedmmsg()
https://quicwg.org/qlog/draft-ietf-quic-qlog-quic-events.html#name-version_information

Comment (nhorman) [reply ssl/quic/quic_port.c:752]: qlog would make sense, but we're not really well positioned to use it at the moment.  qlog seems to be built around the concept of a channel (i.e. QUIC_CHANNELS hold qlog objects and report based on the channel they are connected to).  We could I suppose add a qlog object to a QUIC_PORT, add the appropriate QLOG_EVENT for port events, and add it in here, but I would need to investigate the complexity of such a refactor

Comment (nhorman) [reply ssl/quic/quic_port.c:752]: based on the stand up today, I think we're in consensus that addressing this logging issue should be moved to its own ticket to be addressed, as its a significant amount of additional work to get qlog working at the port level.  Tracked here:
https://github.com/openssl/project/issues/926

Comment (nhorman) [reply ssl/quic/quic_port.c:752]: @t8m add a comment with a TODO (QUIC SERVER) for the retry condition

File: ssl/quic/quic_port.c [hunk 15]
```diff
@@ -620,6 +647,158 @@ static int port_try_handle_stateless_reset(QUIC_PORT *port, const QUIC_URXE *e)
...
+    unsigned char *integrity_tag;
+    int ok;
+
+    /* TODO(QUIC_SERVER): generate proper validation token */
+    memcpy(token.token_buf, "openssltoken", sizeof("openssltoken") - 1);
+
+    token.token_odcid = client_hdr->dst_conn_id;
+    token_len += token.token_odcid.id_len;
+    integrity_tag = (unsigned char *)&token.token_odcid +
+        token.token_odcid.id_len + sizeof(token.token_odcid.id_len);
```
Comment (andrewkdinh) [inline ssl/quic/quic_port.c:696]: Is this overwritten by the call to `ossl_quic_calculate_retry_integrity_tag` on line 719?

File: ssl/quic/quic_port.c [hunk 16]
```diff
@@ -748,6 +748,10 @@ static void port_send_retry(QUIC_PORT *port,
     if (ok == 0)
         return;
 
+    /*
+     * TODO (QUIC SERVER) need to retry this in the event it return EAGAIN
```
Comment (t8m) [inline ssl/quic/quic_port.c:752]: I am sorry for being nitpicking but this needs to be `TODO(QUIC SERVER)` to make it easily searchable. The added space might make it being missed when searching TODOs to resolve.

**Review by t8m — CHANGES_REQUESTED** [review]

**Review by t8m — APPROVED** [review]

**Review by Sashan — APPROVED** [review]
looks good to me. thanks.

**Comment by nhorman** [comment]
rebased onto newly rebased feature/quic-server branch

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the feature branch. I've assumed implicit approval from Neil for the first commit.
